pipeline/coaddition.py

```diff
@@ -0,0 +1,423 @@
+            A matched-filtered score image of the coadded image.
+        """
+        if not all(type(image) == type(images[0]) for image in images):
+            raise ValueError('Not all images are of the same type. ')
+
+        if isinstance(images[0], Image):
+            data = []
+            flags = []
+            weights = []
+            psf_clips = []
+            psf_fwhms = []
+            flux_zps = []
+            
+            for image in images:
+                data.append(image.data)
+                flags.append(image.flags)
+                weights.append(image.weight)
+                psf_clips.append(image.psf.get_clip())
+                psf_fwhms.append(image.psf.fwhm_pixels)
+                flux_zps.append(10 ** (0.4 * image.zp.zp))
+
+        elif isinstance(images[0], np.ndarray):
+            data = images
+        else:
+            raise ValueError('images must be a list of Image objects or 2D arrays. ')
+
+        # pad the PSFs to the same size as the image data
+        psfs = []
+        for array, psf in zip(data, psf_clips):
+            padsize_x1 = int(np.ceil((array.shape[1] - psf.shape[1]) / 2))
+            padsize_x2 = int(np.floor((array.shape[1] - psf.shape[1]) / 2))
+            padsize_y1 = int(np.ceil((array.shape[0] - psf.shape[0]) / 2))
+            padsize_y2 = int(np.floor((array.shape[0] - psf.shape[0]) / 2))
+            psf_pad = np.pad(psf, ((padsize_y1, padsize_y2), (padsize_x1, padsize_x2)))
+            psf_pad /= np.sum(psf_pad)
+            psfs.append(psf_pad)
+
+        # estimate the background if not given
+        if bkg_means is None or bkg_sigmas is None:
+            bkg_means = []
+            bkg_sigmas = []
+            for array in data:
+                bkg, sigma = self._estimate_background(array)
+                bkg_means.append(bkg)
+                bkg_sigmas.append(sigma)
+
+        imcube = np.array(data)
+        flcube = np.array(flags)
+        wtcube = np.array(weights)
+        psfcube = np.array(psfs)
+        bkg_means = np.reshape(np.array(bkg_means), (len(bkg_means), 1, 1))
+        bkg_sigmas = np.reshape(np.array(bkg_sigmas), (len(bkg_sigmas), 1, 1))
+        flux_zps = np.reshape(np.array(flux_zps), (len(flux_zps), 1, 1))
+
+        # subtract the background
+        imcube -= bkg_means
+
+        # make sure to inpaint missing data
+        imcube = self.inpainter.run(imcube, flcube, wtcube)
+
+        if np.sum(np.isnan(imcube)) > 0:
+            raise ValueError('There are still NaNs in the image data after inpainting!')
+
+        # This is where the magic happens
+        outim, psf, score = self._zogy_core(imcube, psfcube, bkg_sigmas, flux_zps)
+
+        # coadd the variance as well
+        varflag = wtcube == 0
+        wtcube2 = wtcube ** 2
+        wtcube2[varflag] = np.nan
+        varmap = 1 / wtcube2
+        varmap = self.inpainter.run(varmap, varflag, wtcube)  # wtcube doesn't do anything, maybe put something else?
```

Comment:
I'm leaving the weights because I don't know what kind of inpainting we will want to do in the future, and I expect images to always have weights anyway. To allow weights=None means figuring out all the ways the inpainting will have to deal with that missing data. I don't think it is worth it.